File: libs/language/src/lib/utils/limits/NodeToPathLimitsConverter.ts
import { BlueNode } from '../../model';
import { PathLimits, PathLimitsBuilder } from './PathLimits';

export class NodeToPathLimitsConverter {
  public static convert(node: BlueNode): PathLimits {
    const builder = new PathLimitsBuilder();
    NodeToPathLimitsConverter.traverseNode(node, '', builder);
    return builder.build();
  }

  private static traverseNode(
    node: BlueNode | undefined,
    currentPath: string,
    builder: PathLimitsBuilder,
  ): void {
    if (!node) {
      return;
    }

    const props = node.getProperties();
    const items = node.getItems();

    const hasProps = props !== undefined && Object.keys(props).length > 0;
    const hasItems = items !== undefined && items.length > 0;

    if (!hasProps && !hasItems) {
      builder.addPath(currentPath === '' ? '/' : currentPath);
      return;
    }

    if (props) {
      for (const [key, value] of Object.entries(props)) {
        const newPath = `${currentPath}/${key}`;
        NodeToPathLimitsConverter.traverseNode(value, newPath, builder);
      }
    }

    if (items) {
      for (let i = 0; i < items.length; i++) {
        const newPath = `${currentPath}/${i}`;
        NodeToPathLimitsConverter.traverseNode(items[i], newPath, builder);
      }
    }
  }
}

File: references/java-sdk/src/main/java/blue/language/utils/limits/NodeToPathLimitsConverter.java
package blue.language.utils.limits;

import blue.language.model.Node;
import blue.language.processor.util.PointerUtils;

import java.util.Map;

public class NodeToPathLimitsConverter {

    public static PathLimits convert(Node node) {
        PathLimits.Builder builder = new PathLimits.Builder();
        traverseNode(node, "/", builder);
        return builder.build();
    }

    private static void traverseNode(Node node, String currentPath, PathLimits.Builder builder) {
        if (node == null) {
            return;
        }

        if ((node.getProperties() == null || node.getProperties().isEmpty()) && node.getItems() == null) {
            builder.addPath(currentPath);
            return;
        }

        if (node.getProperties() != null) {
            for (Map.Entry<String, Node> entry : node.getProperties().entrySet()) {
                String newPath = PointerUtils.appendPointerSegment(currentPath, entry.getKey());
                traverseNode(entry.getValue(), newPath, builder);
            }
        }

        if (node.getItems() != null) {
            for (int i = 0; i < node.getItems().size(); i++) {
                String newPath = PointerUtils.appendEscapedPointerSegment(currentPath, String.valueOf(i));
                traverseNode(node.getItems().get(i), newPath, builder);
            }
        }
    }

}